pinot-query-planner/src/main/java/org/apache/pinot/query/planner/physical/DispatchablePlanContext.java

```diff
@@ -71,17 +88,32 @@ public long getRequestId() {
     return _requestId;
   }
 
+  public PlannerContext getPlannerContext() {
+    return _plannerContext;
+  }
+
+  public PairList<Integer, String> getResultFields() {
+    return _resultFields;
+  }
+
   // Returns all the table names.
   public Set<String> getTableNames() {
     return _tableNames;
   }
 
-  public PairList<Integer, String> getResultFields() {
-    return _resultFields;
+  /// Tracks non-lookup tables queried in leaf stages, which are used to determine the servers to use for intermediate
+  /// stages. Returns `null` if leaf servers are directly used for intermediate stages.
+  @Nullable
+  public Set<String> getNonLookupTables() {
+    return _nonLookupTables;
   }
 
-  public PlannerContext getPlannerContext() {
-    return _plannerContext;
+  /// Tracks servers that are used for leaf stages, which are used to determine the servers to use for intermediate
+  /// stages. Returns `null` if leaf servers are NOT directly used for intermediate stages, and we need to calculate
+  /// the servers to use based on the non-lookup tables.
+  @Nullable
+  public Set<QueryServerInstance> getLeafServerInstances() {
+    return _leafServerInstances;
```

Comment:
Same as before, I think getLeafServerInstances may be used for other things and a developer may blindly assume this method will return not null even if leaf servers are not directly used for intermediate stages.